### PR TITLE
Generate testing log even if solution fails

### LINKED ISSUE #152: Testing log isn't generated in case of failing solution
Any error during testing will prevent a testing log from being generated.
This effectively means `visualize` cannot be used unless you pick a valid time limit beforehand, which makes it effectively useless.

### PR DESCRIPTION
**Pull request by 5K1PY**
Generalized job core for `Job`s running always. `JobManger`s have to generate result even if `JobManager._evaluate` fails.

Fixes #152.

### CODE REVIEW

File: pisek/jobs/jobs.py [hunk 1]
```diff
@@ -100,11 +102,12 @@ def cancel(self) -> None:
...
         self.state = State.canceled
         for item, _ in self.required_by:
-            item.cancel()
+            if not item.run_always:
+                item.cancel()
 
     def _check_prerequisites(self) -> None:
         """Checks if all prerequisites are finished raises error otherwise."""
-        if self.prerequisites > 0:
+        if not self.run_always and self.prerequisites > 0:
```
Comment (SvizelPritula) [inline pisek/jobs/jobs.py:110]: I'll admit I don't know a lot about the inner workings of the pipeline, but isn't this redundant with the changes to `finish`?
`finish` has been changed to decrement `prerequisites` for `run_always` jobs, so I'd assume bypassing the `prerequisites > 0` check is unnecessary?

File: pisek/jobs/parts/testing_log.py [hunk 2]
```diff
@@ -61,12 +65,17 @@ def _evaluate(self) -> None:
                     }
                 )
 
-        if warn_skipped:
-            MSG = "Not all inputs were tested. For testing them use --all-inputs."
+        warn_msg = None
+        if len(solutions) < len(self._env.solutions):
+            warn_msg = "Not all solutions were tested."
+        elif warn_skipped:
+            warn_msg = "Not all inputs were tested. For testing them use --all-inputs."
```
Comment (SvizelPritula) [inline pisek/jobs/parts/testing_log.py:72]: It would be nice to print both warnings, at least if `strict` is not set.

**Comment by SvizelPritula** [comment]
A testing log also doesn't get generated if a solution fails unless `--full` is passed. I haven't managed to get the "Not all solutions were tested." warning to appear.

**Comment by SvizelPritula** [comment]
If you request a testing log when a solution doesn't compile, you gen an exception:

```
Traceback (most recent call last):
  File "/home/dev/.local/bin/pisek", line 8, in <module>
    sys.exit(main_wrapped())
             ^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/__main__.py", line 398, in main_wrapped
    result = main(sys.argv[1:])
             ^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/__main__.py", line 359, in main
    result = test_task(args, solutions=None, target="all")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/utils/pipeline_tools.py", line 77, in g
    res = f(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/__main__.py", line 41, in test_task
    return test_task_path(PATH, **vars(args), **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/__main__.py", line 45, in test_task_path
    return run_pipeline(path, TaskPipeline, solutions=solutions, **env_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/utils/pipeline_tools.py", line 40, in run_pipeline
    return pipeline.run_jobs(Cache(env), env)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/jobs/job_pipeline.py", line 60, in run_jobs
    self.failed |= not self._status_update(env.fork())
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/jobs/job_pipeline.py", line 83, in _status_update
    msg = job_man.finalize()
          ^^^^^^^^^^^^^^^^^^
  File "/home/dev/Code/pisek/pisek/jobs/jobs.py", line 352, in finalize
    self._evaluate()
  File "/home/dev/Code/pisek/pisek/jobs/parts/testing_log.py", line 54, in _evaluate
    for inp, sol_res in data["results"].items():
                        ~~~~^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

File: pisek/jobs/jobs.py (hunk 1 above)
Comment (5K1PY) [reply pisek/jobs/jobs.py:117]: For example when generator fails, testing solutions is canceled and therefore ``finish`` is not called on them. Therefore they do not substract anything from ``self.prerequisites``.

**Comment by 5K1PY** [comment]
> A testing log also doesn't get generated if a solution fails unless `--full` is passed.

That's intentional. I think we should stick to failing fast in all cases unless `--full` is provided.

> I haven't managed to get the "Not all solutions were tested." warning to appear.

One case is if generator fails. I have added also when testing single solution.

File: pisek/jobs/jobs.py (hunk 1 above)
Comment (SvizelPritula) [reply pisek/jobs/jobs.py:117]: In that case, why are the changes in `finish` needed?

**Review by SvizelPritula — APPROVED** [review]
LGTM

**Pull request merged**